### ISSUE TITLE
 Argo .metadata.annotations accessor error in 2.8  

### ISSUE THREAD

**Issue by agaudreault-jive** [argoproj/argo-cd#14336]
**Describe the bug**

After upgrading Argo from 2.5.x -> 2.8.0, I get the following error in my self-managed ArgoCD app. I am using the Argo helm chart 5.34.6, but it does not seem to be the issue. I tried to resync/refresh/hard refresh/invalidate cluster cache/restart redis.
```
ComparisonError: rpc error: code = Unknown desc = Manifest generation error (cached): .metadata.annotations accessor error: <nil> is of the type <nil>, expected map[string]interface{}
```

**Expected behavior**

Successful sync

**Screenshots**

<img width="1490" alt="image" src="https://github.com/argoproj/argo-cd/assets/47184027/7140096a-e92d-4d0c-8190-9eebf90d2d29">

**Version**

```shell
2.8.0
```

**Comment by agaudreault-jive** [comment]
Might come from
```yaml
# Source: argocd/charts/argo-cd/charts/redis-ha/templates/redis-haproxy-service.yaml
apiVersion: v1
kind: Service
metadata:
  name: argocd-redis-ha-haproxy
  namespace: "argocd"
  labels:
    app: redis-ha
    heritage: "Helm"
    release: "argocd"
    chart: redis-ha-4.23.0
    component: argocd-redis-ha-haproxy
  annotations:
spec:
```

**Comment by crenshaw-dev** [comment]
Yep, that matches what I've seen in another case. Not sure what changed causing Argo to no longer handle this well.

**Comment by Raziel9572** [comment]
I'm trying argocd 2.8-rc2 on sandbox environment. I have multiple applications that have this error message. Do I need to set the annotation on every application or is there something I can do on argocd side to fix this for every application ?

**Comment by crenshaw-dev** [comment]
@Raziel9572 for now you'll have to change the manifests. But it is a regression, we need to handle this case in Argo CD.

**Comment by ohauer** [comment]
I think the issue will hit many charts with no values assigned to labels or annotations

- argocd: v2.8.0-rc3 
- chart: bitnami/external-dns 6.20.4

will result in yaml with empty annotation for the ServiceAccount and Deployment unless annotation is set explicit in values.yaml

unluckily I found many many more such charts, so I guess reverting back to argocd v2.7 will be the quickest option for me at the moment + test all charts before on a test cluster with argocd-v2.8.x

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: external-dns
  labels:
  . app.kubernetes.io/name: external-dns
...
  annotations:  <-- per default no values
automountServiceAccountToken: true
```

**Comment by agaudreault-jive** [comment]
@crenshaw-dev Is there a label to identify this bug as a blocker for 2.8?

**Comment by crenshaw-dev** [comment]
@agaudreault-jive I've added it to the roadmap, such is what I've been using to track blockers.

**Comment by agaudreault-jive** [comment]
Seems to be caused by https://github.com/argoproj/argo-cd/pull/13743.
Error is thrown at line https://github.com/argoproj/argo-cd/blob/4ed611c2dd35d4af85b82ee6f82317ea8fcd8fb5/util/kube/kube.go#L25 depending on the ArgoCD tracking method.

**Comment by prein** [comment]
Should I be getting this error in 2.8.1? I understand it was fixed in 2.8.0, correct?
It's the following
```
Failed to load target state: failed to generate manifest for source 1 of 1: rpc error: code = Unknown desc = Manifest generation error (cached): .metadata.labels accessor error: contains non-string key in the map: <nil> is of the type <nil>, expected string
```

**Comment by agaudreault-jive** [comment]
@prein The bug was fixed in the RC before 2.8.0 release.

You can use `helm template` locally to generate the template. Kubernetes requires annotations and labels to be strings. You might have set something in your values.yaml that is not quoted properly and ends up not being interpreted as a string.
